tests/compatibility-test.py

```diff
@@ -128,106 +109,107 @@ def test_kill_head(self):
             raise Exception(f"Nonzero return code {rtn} in test_kill_head()")
 
     def test_ray_serve(self):
+        """Kill GCS process on the head Pod and then test a deployed Ray Serve model."""
         cluster_namespace = "default"
-        docker_client = docker.from_env()
-        container = docker_client.containers.run(ray_image, remove=True, detach=True, stdin_open=True, tty=True,
-                                          network_mode='host', command=["/bin/sh"])
-        # Deploy a model with ray serve
+        headpod = get_head_pod(cluster_namespace)
+        headpod_name = headpod.metadata.name
+
+        # RAY_NAMESPACE is an abstraction in Ray. It is not a Kubernetes namespace.
         ray_namespace = ''.join(random.choices(string.ascii_lowercase, k=10))
-        logger.info(f'namespace: {ray_namespace}')
+        logger.info('namespace: %s', ray_namespace)
 
-        utils.copy_to_container(container, 'tests/scripts', '/usr/local/', 'test_ray_serve_1.py')
-        exit_code, _ = utils.exec_run_container(container, f'python3 /usr/local/test_ray_serve_1.py {ray_namespace}', timeout_sec = 180)
+        # Deploy a Ray Serve model.
+        exit_code = pod_exec_command(headpod_name, cluster_namespace,
+            f" python samples/test_ray_serve_1.py {ray_namespace}",
+            check = False
+        )
 
         if exit_code != 0:
             show_cluster_info(cluster_namespace)
-            raise Exception(f"There was an exception during the execution of test_ray_serve_1.py. The exit code is {exit_code}." +
-                "See above for command output. The output will be printed by the function exec_run_container.")
-
-        # KubeRay only allows at most 1 head pod per RayCluster instance at the same time. In addition,
-        # if we have 0 head pods at this moment, it indicates that the head pod crashes unexpectedly.
-        headpods = utils.get_pod(namespace=cluster_namespace,
-            label_selector='ray.io/node-type=head')
-        assert(len(headpods.items) == 1)
-        old_head_pod = headpods.items[0]
+            raise Exception(
+                f"Fail to execute test_ray_serve_1.py. The exit code is {exit_code}."
```

Comment:
Is it possible to get include the last lines of the driver script output (which should contain a traceback) in the exception for easier debugging? Or is it already easily found somewhere?

Reply:
This boils down to asking whether `subprocess.run(command, shell = True, check = False)` sends stdout and stderr to the right place, which I think it does?

Reply:
Seem ok to me:
```
❯ python -c "import subprocess; subprocess.run('echo hello && blahblahblah', shell=True, check=False)"
hello
/bin/sh: blahblahblah: command not foun
```

Reply:
Ah thanks for checking! I think I was in the mindset of a remote cluster where the driver fails on the cluster and it's hard to see the failure in CI, but here everything is local

Reply:
It would be worth including a comment explaining that the stderr goes to the right place --
I think I asked the same or very similar question when reviewing this code a couple of month's ago.

Reply:
Added a comment to `pod_exec_command`.